src/useMutation.ts

```diff
@@ -32,18 +36,149 @@ export type MutationFn<TData, TVariables> = (
   options?: BaseMutationHookOptions<TData, TVariables>
 ) => Promise<FetchResult<TData>>;
 
-export function useMutation<
-  TData,
-  TVariables = OperationVariables,
-  TCache = object
->(
+export interface ExecutionResult<T = Record<string, any>> {
+  data?: T;
+  extensions?: Record<string, any>;
+  errors?: GraphQLError[];
+}
+
+export interface MutationResult<TData> {
+  called: boolean;
+  data?: TData;
+  error?: ApolloError;
+  hasError: boolean;
+  loading: boolean;
+}
+
+const getInitialState = (): MutationResult<any> => ({
+  called: false,
+  data: undefined,
+  error: undefined,
+  hasError: false,
+  loading: false,
+});
+
+export function useMutation<TData, TVariables = OperationVariables>(
   mutation: DocumentNode,
-  {
-    client: overrideClient,
-    ...baseOptions
-  }: MutationHookOptions<TData, TVariables, TCache> = {}
-): MutationFn<TData, TVariables> {
-  const client = useApolloClient(overrideClient);
-
-  return options => client.mutate({ mutation, ...baseOptions, ...options });
+  baseOptions: MutationHookOptions<TData, TVariables> = {}
```

Comment:
I suggest to extract `client` from `baseOptions` (like before).

Because `client` must not pass into `objToKey ` on [the line below](https://github.com/trojanowski/react-apollo-hooks/pull/135/files#diff-93bc3bb954808aa3a37a149008cd065fR165)

It might lead a bug when use `useMutation(..., { client: customClient })`

```suggestion
  {	
    client: overrideClient,	
    ...baseOptions,
  }: MutationHookOptions<TData, TVariables, TCache> = {}
): MutationFn<TData, TVariables> {
  const client = useApolloClient(overrideClient);
```

Reply:
Actually, I tried it and this is what a bug I got when `client` has a circular field inside.

```
Converting circular structure to JSON
    --> starting at object with constructor 'DefaultClient'
    |     property 'localState' -> object with constructor 'LocalState'
```

Reply:
I had the same issue. Make sure you convert all 
`const toggleTask = useMutation(TOGGLE_TASK_MUTATION);` 
to 
`const [toggleTask] = useMutation(TOGGLE_TASK_MUTATION);`
It fixed the problem for me

Reply:
> I suggest to extract `client` from `baseOptions` (like before).
> 
> Because `client` must not pass into `objToKey ` on [the line below](https://github.com/trojanowski/react-apollo-hooks/pull/135/files#diff-93bc3bb954808aa3a37a149008cd065fR165)
> 
> It might lead a bug when use `useMutation(..., { client: customClient })`

I get exactly this... only way to bypassing it is by setting the context... Do we know if this is being worked?